### PR TITLE
Reversed range

### LINKED ISSUE #10606: Range reverses backwards ranges
```
In [13]: list(range(-1, -10, -1))
Out[13]: [-1, -2, -3, -4, -5, -6, -7, -8, -9]

In [14]: list(Range(-1, -10, -1))
Out[14]: [-9, -8, -7, -6, -5, -4, -3, -2, -1]
```

Range needs to keep the right order so that it is useful for code generation (https://github.com/sympy/sympy/pull/10486).

### PR DESCRIPTION
**Pull request by asmeurer**
@smichr here you go. If you have any suggestions on how to improve (read: simplify) the logic in `Range._intersect`, that would be awesome. I believe the test cases I added cover all the combinations. The infinite range case (particularly infinite start) is nasty because it is handled differently from the others.  If you see anything that you caught that I missed also let me know. 

Fixes #10606.

### CODE REVIEW

File: sympy/printing/pretty/pretty.py [hunk 1]
```diff
@@ -1469,7 +1469,7 @@ def _print_Range(self, s):
         else:
             dots = '...'
 
-        if s.start is S.NegativeInfinity:
+        if s.start in [S.NegativeInfinity, S.Infinity]:
             it = iter(s)
```
Comment (smichr) [inline sympy/printing/pretty/pretty.py]: delete this line

File: sympy/sets/fancysets.py [hunk 2]
```diff
@@ -310,18 +310,62 @@ class Range(Set):
...
+        >>> list(Range(4))
+        [0, 1, 2, 3]
+
+    The step can also be negative. Range(10, 0, -2) represents {10, 8, 6, 4,
+    2}.
+
+        >>> list(Range(10, 0, -2))
+        [10, 8, 6, 4, 2]
+
+    Although this object it a set, it maintains the order of the elements, so
```
Comment (smichr) [inline sympy/sets/fancysets.py]: it a -> is a

File: sympy/sets/fancysets.py [hunk 3]
```diff
@@ -310,18 +310,62 @@ class Range(Set):
...
+    integers, {-oo, ..., -4, -2}.
+
+    The stop value is canonicalized to the largest possible value for the
+    given step, so that equivalent ranges always have the same args. For
+    example, Range(0, 10, 3) and Range(0, 12, 3) both create the same object,
+    representing {0, 3, 6, 9}.
+
+        >>> Range(0, 10, 3) == Range(0, 12, 3)
+        True
+
```
Comment (smichr) [inline sympy/sets/fancysets.py]: I would move the infinite discussion to here, starting:

```
When an range starts with an infinite value, the final value is one step less the provided stop value.
```

Comment (smichr) [inline sympy/sets/fancysets.py]: :pencil2: this is as far as I reviewed.

**Comment by smichr** [comment]
I've got mods in progress in my reversed-range branch. Tests must be added. The commit message of the first commit shows the sort of tests I will add, to be less redundant, however, but sufficient to cover all corner cases.

**Comment by asmeurer** [comment]
Auto-generating tests when both the input and output are finite is fine because we can do the check like https://github.com/sympy/sympy/pull/10660/files#diff-cfbbcfa0f949411825b5f6ef5bd5bcdeR225 to see if the test itself is correct. I don't know how to do that for infinite cases. In the tests I have, each case was verified by me by hand.

**Comment by asmeurer** [comment]
Also, -1 on the Range -> FiniteSet commit. For code generation, we need ranges to stay ranges (EmptySet is an OK exception).

**Comment by asmeurer** [comment]
Also worth reiterating a comment from one of my commit messages

> This changes the behavior of `Range.__iter__` to raise an exception for ranges
with infinite starts, since ranges always iterate in order. See the discussion
on the issue for more information about this. I felt that it was better to
always iterate in order, and catch possible bugs if the range is reversed,
than to iterate backwards in this case. It's worth noting, however, that this
makes it more annoying to determine the values in an infinite range beyond the
two that are printed by the pretty printer. But to fix this, I think we should
add better support for slicing ranges.

(with that being said, I have absolutely no desire to implement `Range.__getitem__(slice)`)

**Comment by smichr** [comment]
> For code generation, we need ranges to stay ranges (EmptySet is an OK exception).

Perhaps it would be better to return the canonical empty `Range(0, 0, 1)`?

**Comment by asmeurer** [comment]
> Perhaps it would be better to return the canonical empty Range(0, 0, 1)?

I don't have strong opinions on this either way, at least yet. I suspect you may be right, though, as otherwise code that uses Range would have to guard against the EmptySet case.

**Comment by smichr** [comment]
I didn't add the Range(0,0,1) case (which I note would also require modifying expectations about equal args) but I am otherwise done with the mods (if tests pass). What do you think about these mods:

- a range that starts with an infinite value is always reversed so the starting value is finite, otherwise it can't be iterated and can only be defined unambiguously for step = +/-1. The only thing one could use such a range for is for intersection but one might as well use S.Integers or friends
- the reversed property is added

btw, I like the use of text after an assertion -- I didn't know one could do that.

**Comment by smichr** [comment]
> makes it more annoying to determine the values in an infinite range

Can you explain this? Did you mean that it complicated the `_intersection` logic?

**Comment by asmeurer** [comment]
> Can you explain this? Did you mean that it complicated the _intersection logic?

I mean just when working with them. If you have something like `Range(-oo, 0, 2)` that prints as `{-oo, ..., -4, -2}`. It's harder to see that the next values here are -6, -8, and so on (in this case it's simple because the step is 2, but it's more complicated when the step is larger). 

> a range that starts with an infinite value is always reversed so the starting value is finite, otherwise it can't be iterated and can only be defined unambiguously for step = +/-1. The only thing one could use such a range for is for intersection but one might as well use S.Integers or friends
the reversed property is added

+1 on the reversed property (I should have added that earlier). We could also probably use it to make things simpler. 

Given that it exists, doesn't it remove the need to reverse ranges that start with infinity? If you want to iterate over them, you can just reverse them first. It does complicate logic, though. 

Do you want to make a PR to my branch, or just merge my branch and make a new PR against master?

**Comment by smichr** [comment]
> Given that it exists, doesn't it remove the need to reverse ranges that start with infinity

**Pro** for inf start: allows an intersection to be ordered like the original range
**Cons**: 

1. can't iterate from it (but we could make it a feature that infinite ranges will always start from the finite end. (As I see it, we might as well since the only other alternative is to raise an error.)
2. logic of other routines have to work with whichever is finite (start of stop); right now things are simplified because the start is always a well behaved reference point for getting on sequence

>  It's harder to see that the next values 

Related, but different: I left a note that said that perhaps we should always make the end value a step of `1` rather than a single `step`; that would always make it easy (maybe?) to see what the last value would be, e.g. `Range(3, 13, 3)` ends at 12 and `Range(12, 2, -1)` ends at 3.



> Do you want to make a PR to my branch, or just merge my branch and make a new PR against master

If you decide for ranges starting from infinity and don't want to change the stop value to be +/-1 of the end value, then we can commit what you have, otherwise I can make a few changes and commit what I have.

**Comment by asmeurer** [comment]
Ranges starting at infinity are already supported so removing them would perhaps be a regression. I don't need them for my personal work, so they're more of a novelty as far as I'm concerned. They might have uses elsewhere, I don't know. I'm not sure if Range will ever be used by solveset. 

> (As I see it, we might as well since the only other alternative is to raise an error.)

In my view an error is better, as it would catch mistakes. 

> logic of other routines have to work with whichever is finite (start of stop); right now things are simplified because the start is always a well behaved reference point for getting on sequence

Sure but my point is that instead of the complicated logic like what I have here, it could be

```py
inf_start = range.start.is_infinite
if inf_start:
    range = range.reversed()
# Do whatever, only needing logic for finite start
if inf_start:
    return result.reversed()
```

> perhaps we should always make the end value a step of 1

No strong opinions on that. 

My preference would be to merge this and you do whatever work you were planning on doing against master, as that would be easiest for me. Assuming you have no objections to what is in here so far. 

Also, my work on ranges with symbolic entries may be postponed, so let me know if you plan to do any work on it, so we don't duplicate efforts. We won't be able to compute much in the general case, but I primarily just want to be able to represent Range(n, m, s) and not have it give any wrong results or exceptions.

**Comment by smichr** [comment]
I left some notes in (perhaps an outdated diff) at https://github.com/asmeurer/sympy/pull/74/files. If it passes and you're ok with it, then apply to this and commit. If tests fail there, go ahead and commit here and I will work it out later.

**Pull request closed without merging**